storage/src/persistent/commit_log/mod.rs

```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
+    ReadError { location: Location },
+    #[fail(display = "Failed to read record data corrupted")]
+    CorruptData,
+    #[fail(display = "Tezedge CommitLog error: {:#?}", error)]
+    TezedgeCommitLogError { error: TezedgeCommitLogError },
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Index {
+    pub position: u64,
+    pub compressed_data_length: u64,
+    pub uncompressed_data_length: u64,
+}
+
+impl Index {
+    fn from_buf(buf: &[u8]) -> Result<Self, TezedgeCommitLogError> {
+        if buf.len() != TH_LENGTH {
+            return Err(TezedgeCommitLogError::IndexLengthError);
+        }
+        let mut buf = buf.to_vec();
+        let position_raw_bytes: Vec<_> = buf.drain(..8).collect();
+        let compressed_data_length_raw_bytes: Vec<_> = buf.drain(..8).collect();
+        let uncompressed_data_length_raw_bytes: Vec<_> = buf.drain(..).collect();
+        let position = u64::from_be_bytes(
+            position_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        let compressed_data_length = u64::from_be_bytes(
+            compressed_data_length_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        let uncompressed_data_length = u64::from_be_bytes(
+            uncompressed_data_length_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        Ok(Self {
+            position,
+            compressed_data_length,
+            uncompressed_data_length,
+        })
+    }
+    fn new(position: u64, compressed_data_length: u64, uncompressed_data_length: u64) -> Self {
+        Self {
+            position,
+            compressed_data_length,
+            uncompressed_data_length,
+        }
+    }
+    fn to_vec(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        for b in &self.position.to_be_bytes() {
+            buf.push(*b)
+        }
+        for b in &self.compressed_data_length.to_be_bytes() {
+            buf.push(*b)
+        }
+        for b in &self.uncompressed_data_length.to_be_bytes() {
+            buf.push(*b)
+        }
+        return buf;
+    }
+}
+
+pub struct CommitLog {
+    path: PathBuf,
+}
+
+impl CommitLog {
+    pub fn new<P: AsRef<Path>>(log_dir: P) -> Self {
+        Self {
+            path: log_dir.as_ref().to_path_buf(),
+        }
+    }
+    #[inline]
+    pub fn append_msg<B: AsRef<[u8]>>(&mut self, payload: B) -> Result<u64, TezedgeCommitLogError> {
+        let mut writer = Writer::new(&self.path)?;
+        let offset = writer.write(payload.as_ref())?;
+        Ok(offset)
+    }
+
+    #[inline]
+    pub fn read(&self, from: usize, limit: usize) -> Result<MessageSet, TezedgeCommitLogError> {
+        let mut reader = Reader::new(&self.path)?;
```

Comment:
@mambisi 
do we need to create new instances `Reader::new/Writer:new`, all the time? it does lots of things in `new`

Reply:
No it doesn't but the BufWriter need to be drop for changes to take effect, we can make on reader, i will try it out

Reply:
@mambisi isn't calling `flush()` on the writer enough? The drop just falls `flush_buf` which is also called by `flush` https://doc.rust-lang.org/src/std/io/buffered/bufwriter.rs.html#414-421

Reply:
I think it will work